File: software/mosek/7/tools/examples/java/callback.java
package com.mosek.example;
//
//   Copyright: Copyright (c) MOSEK ApS, Denmark. All rights reserved.
//
//   File:      callback.java
//
//   Purpose:   To demonstrate how to use the progress 
//              callback. 
//
//              Use this script as follows:
//              callback psim  25fv47.mps
//              callback dsim  25fv47.mps
//              callback intpnt 25fv47.mps
//
//              The first argument tells which optimizer to use
//              i.e. psim is primal simplex, dsim is dual simplex
//              and intpnt is interior-point. 
//

import mosek.*;
import java.util.Formatter;

public class callback
{
    private static Progress makeUserCallback(final double maxtime)
    {
      return new Progress() 
      {
          public int progress(int code, 
                              double[] douinf,
                              int[]    intinf,
                              long[]   lintinf)
          {
            Env.callbackcode caller = Env.callbackcode.fromValue(code);
            double opttime = 0.0;
            int itrn;
            double pobj, dobj, stime;

            Formatter f = new Formatter(System.out);
            switch (caller)
            {
              case begin_intpnt:
                  f.format("Starting interior-point optimizer");
              case intpnt:
                  itrn    = intinf[Env.iinfitem.intpnt_iter.value      ];
                  pobj    = douinf[Env.dinfitem.intpnt_primal_obj.value];
                  dobj    = douinf[Env.dinfitem.intpnt_dual_obj.value  ];
                  stime   = douinf[Env.dinfitem.intpnt_time.value      ];
                  opttime = douinf[Env.dinfitem.optimizer_time.value   ];

                  f.format("Iterations: %-3d",itrn);
                  f.format("Time: %6.2f(%.2f) ",opttime,stime);
                  f.format("Primal obj.: %-18.6e  Dual obj.: %-18.6e",pobj,dobj);
              case end_intpnt:
                  f.format("Interior-point optimizer finished.");
              case begin_primal_simplex:
                  f.format("Primal simplex optimizer started.");
              case update_primal_simplex:
                  itrn    = intinf[Env.iinfitem.sim_primal_iter.value  ];
                  pobj    = douinf[Env.dinfitem.sim_obj.value          ];
                  stime   = douinf[Env.dinfitem.sim_time.value         ];
                  opttime = douinf[Env.dinfitem.optimizer_time.value   ];
            
                  f.format("Iterations: %-3d", itrn);
                  f.format("  Elapsed time: %6.2f(%.2f)",opttime,stime);
                  f.format("Obj.: %-18.6e", pobj );
              case end_primal_simplex:
                  f.format("Primal simplex optimizer finished.");
              case begin_dual_simplex:
                  f.format("Dual simplex optimizer started.");
              case update_dual_simplex:
                  itrn    = intinf[Env.iinfitem.sim_dual_iter.value    ];
                  pobj    = douinf[Env.dinfitem.sim_obj.value          ];
                  stime   = douinf[Env.dinfitem.sim_time.value         ];
                  opttime = douinf[Env.dinfitem.optimizer_time.value   ];
                  f.format("Iterations: %-3d", itrn);
                  f.format("  Elapsed time: %6.2f(%.2f)",opttime,stime);
                  f.format("Obj.: %-18.6e", pobj);
              case end_dual_simplex:
                  f.format("Dual simplex optimizer finished.");
              case begin_bi:
                  f.format("Basis identification started.");
              case end_bi:
                  f.format("Basis identification finished.");
              default:
            }
            System.out.flush();
            if (opttime >= maxtime)
              // mosek is spending too much time. Terminate it.
              return 1;
            
            return 0;
          }
      };
    }

    public static void main(String[] args)
    {
      if (args.length < 2)
      {
        for (int i=0;i<args.length;++i)
        {
          System.out.println(args[i]);
        }
        
        System.out.println("Too few input arguments. Syntax:");
        System.out.println("\tcallback.py psim inputfile");
        System.out.println("\tcallback.py dsim inputfile");
        System.out.println("\tcallback.py intpnt inputfile");
      }
      else
      {
        Env env = null;
        Task task = null;
        try
        {
          String filename = args[1];
          env = new Env();
          task = new Task(env,0,0);

          task.set_Stream(
            mosek.Env.streamtype.log,
            new mosek.Stream() 
            { public void stream(String msg) { System.out.print(msg); }});

          task.readdata(filename);                    

          if   (args[0] == "psim")
            task.putintparam(Env.iparam.optimizer,Env.optimizertype.primal_simplex.value);
          else if (args[0] == "dsim")            
            task.putintparam(Env.iparam.optimizer,Env.optimizertype.dual_simplex.value);
          else if (args[0] == "intpnt")
            task.putintparam(Env.iparam.optimizer,Env.optimizertype.intpnt.value);

          // Turn all MOSEK logging off (note that errors and other messages 
          // are still sent through the log stream)
          task.putintparam(Env.iparam.log, 0);
        
          task.set_Progress(makeUserCallback(3600));
          
          task.optimize();
          task.putintparam(Env.iparam.log, 1);
          task.solutionsummary(Env.streamtype.msg);

        }
        catch (mosek.Exception e)
        {
          System.out.println ("An error/warning was encountered");
          System.out.println (e.toString());
          throw e;
        }
        finally 
        {
          if (task != null)
            task.dispose();
          if (env != null)
            env.dispose();
        }
      }
   }
}

File: software/mosek/7/tools/examples/python/callback.py
##
#   Copyright: Copyright (c) MOSEK ApS, Denmark. All rights reserved.
#
#   File:      callback.py
#
#   Purpose:   To demonstrate how to use the progress 
#              callback. 
#
#              Use this script as follows:
#              callback.py psim  25fv47.mps
#              callback.py dsim  25fv47.mps
#              callback.py intpnt 25fv47.mps
#
#              The first argument tells which optimizer to use
#              i.e. psim is primal simplex, dsim is dual simplex
#              and intpnt is interior-point. 
##
from __future__ import with_statement

import sys

import mosek
from mosek import *


def makeUserCallback(maxtime):
    def userCallback(caller,
                     douinf,
                     intinf,
                     lintinf):
        opttime = 0.0

        if   caller == callbackcode.begin_intpnt:
            print ("Starting interior-point optimizer")
        elif caller == callbackcode.intpnt:
            itrn    = intinf[iinfitem.intpnt_iter      ]
            pobj    = douinf[dinfitem.intpnt_primal_obj]
            dobj    = douinf[dinfitem.intpnt_dual_obj  ]
            stime   = douinf[dinfitem.intpnt_time      ]
            opttime = douinf[dinfitem.optimizer_time   ]

            print ("Iterations: %-3d" % itrn)
            print ("  Elapsed time: %6.2f(%.2f) " % (opttime,stime))
            print ("  Primal obj.: %-18.6e  Dual obj.: %-18.6e" % (pobj,dobj))
        elif caller == callbackcode.end_intpnt:
            print ("Interior-point optimizer finished.")
        elif caller == callbackcode.begin_primal_simplex:
            print ("Primal simplex optimizer started.")
        elif caller == callbackcode.update_primal_simplex:
            itrn    = intinf[iinfitem.sim_primal_iter  ]
            pobj    = douinf[dinfitem.sim_obj          ]
            stime   = douinf[dinfitem.sim_time         ]
            opttime = douinf[dinfitem.optimizer_time   ]
            
            print ("Iterations: %-3d" % itrn)
            print ("  Elapsed time: %6.2f(%.2f)" % (opttime,stime))
            print ("  Obj.: %-18.6e" % pobj )
        elif caller == callbackcode.end_primal_simplex:
            print ("Primal simplex optimizer finished.")
        elif caller == callbackcode.begin_dual_simplex:
            print ("Dual simplex optimizer started.")
        elif caller == callbackcode.update_dual_simplex:
            itrn    = intinf[iinfitem.sim_dual_iter    ]
            pobj    = douinf[dinfitem.sim_obj          ]
            stime   = douinf[dinfitem.sim_time         ]
            opttime = douinf[dinfitem.optimizer_time   ]
            print ("Iterations: %-3d" % itrn)
            print ("  Elapsed time: %6.2f(%.2f)" % (opttime,stime))
            print ("  Obj.: %-18.6e" % pobj)
        elif caller == callbackcode.end_dual_simplex:
            print ("Dual simplex optimizer finished.")
        elif caller == callbackcode.begin_bi:
            print ("Basis identification started.")
        elif caller == callbackcode.end_bi:
            print ("Basis identification finished.")
        else:
            pass

        if opttime >= maxtime:
            # mosek is spending too much time. Terminate it.
            return 1
      
        return 0
    return userCallback

def msgPrinter(msg):
    sys.stdout.write(msg)
    sys.stdout.flush()

def main(args):

  if len(args) < 3:
      print ("Too few input arguments. Syntax:")
      print ("\tcallback.py psim inputfile")
      print ("\tcallback.py dsim inputfile")
      print ("\tcallback.py intpnt inputfile")
      return 
  
  with mosek.Env() as env:
      with mosek.Task(env) as task:
          filename = args[2]
          task.readdata(filename)

          task.set_Stream(streamtype.log, msgPrinter)

          if   args[1] == 'psim':
              task.putintparam(iparam.optimizer,optimizertype.primal_simplex)
          elif args[1] == "dsim":
              task.putintparam(iparam.optimizer,optimizertype.dual_simplex)
          elif args[1] == "intpnt":
              task.putintparam(iparam.optimizer,optimizertype.intpnt)

          # Turn all MOSEK logging off (note that errors and other messages 
          # are still sent through the log stream)
          task.putintparam(iparam.log, 0)
        
          usercallback = makeUserCallback(maxtime = 3600)
          task.set_Progress(usercallback)

          task.optimize()

          task.solutionsummary(streamtype.msg)

if __name__ == '__main__':
    main(sys.argv)
